src/tensor_operations/matrix_decomposition.jl

```diff
@@ -369,48 +369,142 @@ function noinds_error_message(decomp::String)
+  # Make a dummy index with dim=1 and incorporate into A so the Lis & Ris can never
+  # be empty.  A essentially becomes 1D after collection.
+  A, vαl, vαr, Lis, Ris = add_trivial_index(A, Lis, Ris)
+
+  #
+  #  Use combiners to render A down to a rank 2 tensor ready matrix QR routine.
+  #
+  CL, CR = combiner(Lis...), combiner(Ris...)
+  cL, cR = combinedind(CL), combinedind(CR)
+  AC = A * CR * CL
+  #
+  #  Make sure we don't accidentally pass the transpose into the matrix qr routine.
+  #
+  AC = permute(AC, cL, cR; allow_alias=true)
+  # qr the matrix.
+  QT, RT = qr(tensor(AC); kwargs...)
+  #
+  #  Undo the combine oepration, to recover all tensor indices.
+  #
+  Q, R = itensor(QT) * dag(CL), itensor(RT) * dag(CR)
+
+  # Remove dummy indices.  No-op if vαl and vαr are Nothing
+  Q, R = remove_trivial_index(Q, R, vαl, vαr)
+  #
+  # fix up the tag name for the index between Q and R.
+  #  
   q = commonind(Q, R)
-  settags!(Q, tags, q)
-  settags!(R, tags, q)
-  q = settags(q, tags)
 
-  if isempty(Lis_original)
-    Q *= dag(vLα)
-  end
-  if isempty(Ris_original)
-    R *= dag(vRα)
-  end
+  Q = settags(Q, qtag, q)
+  R = settags(R, qtag, q)
+  q = settags(q, qtag)
 
   return Q, R, q
 end
 
+function rq(A::ITensor, Linds...; tags=ts"Link,rq", kwargs...)
+  qtag = TagSet(tags) #tag for new index between Q and R
+  Lis = commoninds(A, indices(Linds...))
+  Ris = uniqueinds(A, Lis)
+  # make a dummy index with dim=1 and incorporate into A so the Lis & Ris can never
+  # be empty.  A essentially becomes 1D after collection.
+  A, vαl, vαr, Lis, Ris = add_trivial_index(A, Lis, Ris)
+
+  #
+  #  Use combiners to render A down to a rank 2 tensor ready matrix QR routine.
+  #
+  CL, CR = combiner(Lis...), combiner(Ris...)
+  cL, cR = combinedind(CL), combinedind(CR)
+  AC = A * CR * CL
+  #
+  #  Make sure we don't accidentally pass the transpose into the matrix qr routine.
+  #
+  AC = permute(AC, cL, cR; allow_alias=true)
+  # qr the matrix.
+  RT, QT = NDTensors.rq(tensor(AC); kwargs...)
+  #
+  #  Undo the combine oepration, to recover all tensor indices.
+  #
+  R, Q = itensor(RT) * dag(CL), itensor(QT) * dag(CR)
+
+  # Conditionally remove dummy indices.
+  R, Q = remove_trivial_index(R, Q, vαl, vαr)
+  #
+  # fix up the tag name for the index between Q and R.
+  #  
+  q = commonind(Q, R)
+  Q = settags(Q, qtag, q)
+  R = settags(R, qtag, q)
+  q = settags(q, qtag)
+
+  return R, Q, q
+end
+
+# lq is exported by the LinearAlgebra module so we need acknowledge that to avoid
+# intermitent run time errors.
+function lq(A::ITensor, Linds...; tags=ts"Link,lq", kwargs...)
+  Q, L, q = qr(A, uniqueinds(A, Linds...); kwargs...)
+  #
+  # fix up the tag name for the index between Q and R.
+  #  
+  qtag = TagSet(tags)
+  Q = settags(Q, qtag, q)
+  L = settags(L, qtag, q)
+  q = settags(q, qtag)
+
+  return L, Q, q
+end
+
+function ql(A::ITensor, Linds...; tags=ts"Link,ql", kwargs...)
+  Q, L, q = rq(A, uniqueinds(A, Linds...); kwargs...)
```

Comment:
Shouldn't this output `R, Q, q`?

Reply:
(Or `L, Q, q`, I guess the name doesn't matter and it makes sense to name it `L`, but the question stands that isn't the 2nd output of `rq` the orthogonal matrix?).

Reply:
Yes I am all for re-factoring and minimizing the amount of code that must be maintained.  I will have a look.  I noticed the Q,L swap this morning I will fix that.

Reply:
That was a useful refactor.  A lot of code reduction.  The qr_positive and ql_positive routines can probably be refactored into one, with a slight reduction in performance (*-1 for wholes rows instead of partial rows).

Reply:
Found a bug in ql_positive not captured by the current unit tests.  Fix and new unit tests in process.

Reply:
If you agree refactoring ql_positive and qr_positive into one generic routine is not useful, then this is done.

Reply:
Are you asking if you should refactor `qr_positive(M::AbstractMatrix)` and `ql_positive(M::AbstractMatrix)` into one generic function? I think it could make sense to make a single generic `qx_positive` function, if they are actually implemented the same way (i.e. analyzing the sign of the diagonal entries and then multiplying those signs into `Q`).

However, I'm a bit confused why they look different, it seems like `ql_positive` has some extra logic for the case `nc > nr` that `qr_positive` doesn't have? That already seems like a good reason for making them one function, assuming that logic should be shared by both functions.

Additionally, it doesn't seem like the code as it is written would handle the complex case properly. In general for complex number there would be a phase along the diagonals and the diagonal entries of `R`/`L` should get transformed into the absolute value of the diagonal entry. Julia's `sign` function generically handles that:
```julia
julia> sign(1 + 2im)
0.4472135954999579 + 0.8944271909999159im

julia> sign(-1)
-1
```

Reply:
issue #1:
The difference has to do with where the diagonal is for rectangular R and L matricies.  First off for QR/QL, if they rectangular then nc>nr.  Under these conditions for R the diagonal is right where you expect it.  But for L the diagonal is shifted over to the right by dc=nc-nr.  So L is different. It should be easy to make this generic, but the code might be a little harder to comprehend.
issue #2
When we change the sign of a row, R we can ignore all zeros to the left of the diag, and for L we can ignore the zeros to the right of the diag.  A simple generic version with flip the sign of the whole row which is slightly less efficient.  I doubt we would have an easy time measuring the difference though, the time the lapack would probably swamp the difference.
issues #3
complex handling:  Let me investigate.

Reply:
I see, thanks for the summary. It's a bit more complicated than I was imagining and the functions aren't so complicated so let's just leave them as they are (but we should make sure we handle the complex case properly).

Reply:
It turns the complex case is already handled properly, it seems lapack is doing that for us.  It is returning all reals on the diagonals of R and L.
I have beefed up the unit tests at the NDTensor level in include all combinations:

1. Nrows>Ncols, Ncols>Nrows
2. qr,ql
3. Float64, ComplexF64, Float32, ComplexF32
4. positive in [false,true]
5. If positive=true, check all(real(diagX).>=0.0) && all(imag(diagX).==0.0)

So maybe we should leave qr_positive and ql_positive as is, but I can add some comments about the complex case.
I will wait for confirmation (or lack thereof :) before pushing the unit test upgrade to avoid triggering a full CI run.

Reply:
I would rather not rely on LAPACK returning reals on the diagonals of `R` and `L`. My guess is that isn't guaranteed by the API (for example, in https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-0/gerqf.html#API I don't see anything mentioned about guarantees of constraints on the diagonal values of `R`), and could depend on the particular LAPACK backend, architecture, version, etc.

It should be a simple change to make it more general, i.e. changing:
```julia
    if real(R[c, c]) < 0.0
      R[c, c:end] *= -1 #only fip non-zero portion of the row.
      Q[:, c] *= -1
    end
```
to:
```julia
    sign_Rc = sign(R[c, c])
    if !isone(sign_Rc)
      R[c, c:end] *= conj(sign_Rc) #only fip non-zero portion of the row.
      Q[:, c] *= sign_Rc
    end
```
This should be exactly the same for the real diagonal case but then also cover any complex diagonal cases that may arise.

Reply:
Excellent point ... I forgot about all those vendor specific (partial) dropins for lapack floating around out there.

Reply:
Done